### DISCUSSION TITLE
Issues Loading and Testing WebAssembly Modules in `vitest`

### DISCUSSION THREAD

**Discussion by bennobuilder** [vitest-dev/vitest#4283] [Q&A]
I'm working on a project that involves WebAssembly (WASM) modules compiled from Rust. While these modules work flawlessly in my Next.js & NodeJs (bundled using Rollup to `dist` folder) application, I'm encountering challenges when trying to test them using `vitest`.

Here's my test setup:

```javascript
import { greetRust, initWasm } from '../../dist/esm/index.js'; // Load way 2
import { greetRust, initWasm } from '..'; // Load way 1

describe('myWasmModule', () => {
    beforeEach(async () => {
        await initWasm();
    });

    it('should execute wasm function without errors', async () => {
        const result = greetRust();
        expect(result).toBe("Hello Rust 7");
    });
});
```

I've attempted to load the WASM module in two ways:

1. **Directly from TypeScript**:
   
   By importing from `'..'`, I receive the following error:
   
   ```
   Error: "ESM integration proposal for Wasm" is not supported currently. Use vite-plugin-wasm or other community plugins to handle this. Alternatively, you can use `.wasm?init` or `.wasm?url`.
   ```
Are there any 

2. **From the `dist` directory**:
   
   By importing from `'../../dist/esm/index.js'`, I get a `RuntimeError: unreachable` error during test execution.

Given the above, I have several questions:

1. Are there specific configurations or plugins I need to set up for `vitest` to handle WASM modules properly?
2. Are there best practices or recommended approaches for testing WASM modules in `vitest`?

Any guidance would be appreciated :) 

Thanks a lot

**Context**

`index.ts`
```
import init, { add, cfgTest } from '@rust/dyn-dtom';
import wasm from '@rust/dyn-dtom/bg.wasm';

export async function initWasm(): Promise<void> {
	// https://www.npmjs.com/package/@rollup/plugin-wasm#using-with-wasm-bindgen-and-wasm-pack
	// @ts-expect-error - "wasm" needs to be loaded async
	const wasmInstance: unknown = await wasm();
	await init(wasmInstance);
}

export function greetRust(): string {
	cfgTest();
	return `Hello Rust ${add(3, 4)}`;
}
```

Project Structure:
![image](https://github.com/vitest-dev/vitest/assets/57860196/8c44269c-32b8-484c-96bc-0fbb0b62dd4c)

Note: that rn the WASM is inlined in `bg.wasm.js` as Base64 string as I couldn't figure out how to make NextJs accept `.wasm` files as resources when the WASM is located in a `npm` package. See: https://github.com/vercel/next.js/discussions/56634

**Comment by janosh** [comment]
@sheremet-va @antfu @AriPerkkio could this be converted from a discussion to an issue aimed at adding docs on how to `vitest` code that imports WASM binaries? the 20 upvotes on this discussion signal there's some interest in this

**Reply by hi-ogawa** [reply]
Please feel free to create a new issue with a reproduction if you have a concrete example. Or if you already have a solution and proposal for doc update, that's also welcome.

**Reply by r20** [reply]
What finally helped me was @hi-ogawa 's example source code linked at https://github.com/vitest-dev/vitest/issues/2910#issuecomment-1905313636
